cartoframes/viz/kuviz.py

```diff
@@ -45,3 +52,50 @@ def _validate_carto_kuviz(carto_kuviz):
 def _create_carto_kuviz(context, html, name, password=None):
     km = KuvizManager(context.auth_client)
     return km.create(html=html, name=name, password=password)
+
+
+class KuvizPublisher(object):
+    def __init__(self, vmap, context=None):
+        self._layers = deepcopy(vmap.layers)
+        self._context = context
+
+    def set_context(self, context=None):
+        from cartoframes.auth import _default_context
+        self._context = context or _default_context
+
+    def publish(self, html, name, password=None):
+        return Kuviz.create(context=self._context, html=html, name=name, password=password)
+
+    def is_sync(self):
+        syncs = [layer.source.dataset.get_is_saved_in_carto() for layer in self._layers]
+        return False not in syncs
+
+    def get_layers(self, maps_api_key='default_public'):
+        for layer in self._layers:
+            layer.source.dataset.set_context(self._context)
+
+            layer.source.credentials = {
+                'username': self._context.creds.username(),
+                'api_key': maps_api_key,
+                'base_url': self._context.creds.base_url()
+            }
+
+        return self._layers
+
+    def sync_layers(self, table_name, context=None):
+        for idx, layer in enumerate(self._layers):
+            table_name = normalize_name("{name}_{idx}".format(name=table_name, idx=idx + 1))
+
+            from cartoframes.auth import _default_context
+            dataset_context = context or layer.source.dataset.get_context() or _default_context
+
+            self._sync_layer(layer, table_name, dataset_context)
+
+    def _sync_layer(self, layer, table_name, context):
+        if not layer.source.dataset.get_is_saved_in_carto():
+            layer.source.dataset.upload(table_name=table_name, context=context)
+            layer.source = Source(table_name, context=context)
+            warn('Table `{}` created. In order to publish the map, you will need to create a new Regular API '
```

Comment:
It'd be pretty sweet if we could help them do that automagically (and explicitly) through cartoframes so that we take away the disruption to the workflow that getting a new key entails.

Related: https://github.com/CartoDB/cartoframes/issues/542

Reply:
We don't have Auth API implemented in carto-python yet, but when we have it, we will be able to do it automagically